src/common/base/EnumArray.h

```diff
@@ -0,0 +1,249 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#ifndef COMMON_BASE_ENUMARRAY_H_
+#define COMMON_BASE_ENUMARRAY_H_
+
+#include "base/Base.h"
+
+namespace nebula {
+
+template<bool C, typename Then, typename Else>
+class _IF {};
+
+template<typename Then, typename Else>
+class _IF<true, Then, Else> {
+public:
+    typedef Then reType;
+};
+
+template<typename Then, typename Else>
+class _IF<false, Then, Else> {
+public:
+    typedef Else reType;
+};
+
+template<template<class> class Condition, class Statement>
+class _WHILE {
+    template<typename T>
+    class _STOP {
+    public:
+        typedef T reType;
+    };
+
+public:
+    typedef typename
+        _IF<Condition<Statement>::ret,
+        _WHILE<Condition, typename Statement::Next>,
+        _STOP<Statement>>::reType::reType
+    reType;
+};
+
+template<size_t N>
+class NextPowerTwo {
+    template<typename Stat>
+    class Cond {
+    public:
+        enum {
+            ret = (Stat::number != 0),
+        };
+    };
+
+    template<size_t n, uint8_t cnt>
+    class Stat {
+    public:
+        enum {
+            number = n >> 1,
+            count = cnt + 1,
+        };
+
+        typedef Stat<number, count> Next;
+    };
+
+    template<size_t n>
+    class isPowerTwo {
+    public:
+        enum {
+            ret = (n != 0) && !(n & (n - 1)),
```

Comment:
this mean ret is true just when n=1,2,4,8,16... like 0x100000...0000?

Reply:
Ha,  it just handles the unsigned integer.